### DISCUSSION TITLE
api directory with its own package.json causing problems

### DISCUSSION THREAD

**Discussion by elijahcarrel** [vercel/vercel#4540] [Help] [answered]
Hello! I have [an app](https://github.com/elijahcarrel/inbox-comics) with a Next.js front-end and a node back-end.

Currently, I do this via having two completely separate vercel projects (one for the front end and one for the back end) with different deployments, different domains that communicate to each other via CORS. This is not ideal. So, I am trying to consolidate into one vercel project that makes use of the `api` directory for the back-end functions.

I am almost able to get this to work. I have a `package.json` in the root directory for the next.js UI and a second `package.json` in the `api` directory for the node backend. My `vercel.json` file looks something like this:
```
{
  "version": 2,
  "builds": [
    { "src": "package.json", "use": "@now/next" },
    { "src": "api/src/index.ts", "use": "@now/node" }
  ],
  "routes": [
    { "src": "/api/(.*)", "dest": "/api/src/index.ts" }
  ]
}
```

My [WIP](https://github.com/elijahcarrel/inbox-comics/tree/one-deployment) works with `vercel dev`. However, as soon as I deploy to a preview/prod deployment, it fails at the build step with webpack errors of the form `Cannot find module 'xyz' or its corresponding type declarations.` for every import in a javascript file in the `api` directory that references a package `xyz` that is listed as a dependency in `[git root]/api/package.json` but not in `[git root]/package.json`. My issue appears quite similar to https://github.com/vercel/vercel/issues/2840, but I am not sure how to resolve it. I would certainly prefer not to have to list all my API dependencies as dependencies in `[git root]/package.json` as well, since that would seem to imply that they are dependencies of the UI when they are strictly used by the server. The error is also especially bizarre because earlier logs indicate that `api/node_modules` are correctly getting installed. I cannot even tell if these errors are occurring when Vercel is trying to build the backend, or if they are occurring when when Vercel is trying to build the front-end (and they are getting incorrectly looped in by webpack for some reason).

A couple questions:
(1) How/when are the `api` functions built? Are they built during the build step at all, or does that not happen until someone actually tries to go to the website?
(2) Does anyone have ideas for what the actual issue is, and how I could go about resolving it? E.g. should I write a custom webpack config in the root directory explicitly excluding the `api` directory, because maybe they're incorrectly getting included in my webpack tree for the UI? Should I write a custom webpack config in the API directory telling webpack where to resolve `node_modules` to, because it seems to be failing at that part? 

If this is not the appropriate forum for this sort of question, please let me know where a more appropriate forum might be. Thanks.

**Comment by user72356** [comment]
I have the exact same problem and questions as you. Did you ever find answers? Thank you!

**Comment by amyegan** [comment, marked as answer]
It's best to think of `/api` as part of the larger app rather than as a separate backend. Serverless Functions are created during the build, and there is only one build step per project. The root `package.json` is the only source of truth during that build step. 

If you prefer to keep the UI and Serverless Functions separate, you should explore [rewrites](https://vercel.com/docs/project-configuration#project-configuration/rewrites) and [Build Output API (v3)](https://vercel.com/docs/build-output-api/v3#).
